### PR TITLE
adding uploading media locally script

### PR DESCRIPTION
**Pull request by kasra-r77**
Adding upload script in python for handling local uploads to Studio API:
- Added utils folder to put generic methods there for future reuse in other scripts
- Added `.gitignore`
- Updated README

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,34 @@ In this repository, we collect example scripts that can be useful for the team a
 
 ## Prerequisites
 
-TBD
+To run python scripts inside `examples` directory you should first run the installer script:
+
+```bash
+$ bin/install
```
Comment (zoltan-sebestyen-instructure) [inline README.md:10]: imho this `bin/run` and `bin/install` are overkill. For the media downloaded I added the following and the users were happy with it, even though I forgot to include the `source venv/bin/activate` as a pre-requisite to run the script.

install: 

> - install a python 3 interpreter (if not installed already)
> - create a directory where you would run the script from.
> - copy the script to the directory.
> - install the ‘requests’ library by running ‘pip install -m requirements.txt’ from command line in that directory.

File: bin/install [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
Comment (zoltan-sebestyen-instructure) [inline bin/install:1]: see above, I don't think we need this script.

File: bin/run [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
Comment (zoltan-sebestyen-instructure) [inline bin/run:1]: see above, I don't think we need this script.

File: examples/upload-media-from-local/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+This is a script to handle uploading media locally to Studio.
+
+---
+
+## How to use
+
+For using this script we should run it through runner script in `bin` folder:
+
+```bash
+$ bin/run examples/upload-media-from-local/upload.py <arguments>
```
Comment (zoltan-sebestyen-instructure) [inline examples/upload-media-from-local/README.md:12]: same here

**Review by zoltan-sebestyen-instructure — CHANGES_REQUESTED** [review]
I would add the jira ticket it to the commit message.

File: README.md [hunk 5]
```diff
@@ -4,7 +4,34 @@ In this repository, we collect example scripts that can be useful for the team a
...
+```
+
+This will handle installing `venv` and requierd packages.
```
Comment (zoltan-sebestyen-instructure) [inline README.md]: I would mention the `.config` file here.

File: README.md [hunk 6]
```diff
@@ -4,7 +4,34 @@ In this repository, we collect example scripts that can be useful for the team a
...
+
+After the installtion, you can run the scripts using `bin/run` script:
+
+```bash
+$ bin/run <path_to_script_file> <arguments>
+```
+
+### Authorization
+
+In order to authorize the requests, please add your credentails in a `JSON` format into the `.config` file:
```
Comment (zoltan-sebestyen-instructure) [inline README.md]: Shouldn't we mention https://community.canvaslms.com/t5/The-Product-Blog/Connecting-Studio-OAuth-via-Postman/ba-p/259739 ?

File: README.md [hunk 7]
```diff
@@ -17,9 +44,10 @@ Each script needs to have its own subdirectory under the `examples` directory an
 i.e. if there's a script to download the captions of a media, the layout should look like:
 
 ```
+
 examples/
```
Comment (zoltan-sebestyen-instructure) [inline README.md:48]: I would revert this change as without the indentation it's misleading.

File: bin/install [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && cd .. && pwd)"
+
+if [ ! -d "${__dir}/venv" ]; then
+  python3 -m venv "${__dir}/venv"
+fi
+
+source "${__dir}/venv/bin/activate"
+
+if [ ! -d "${__dir}/.config" ]; then
```
Comment (zoltan-sebestyen-instructure) [inline bin/install:12]: I believe what you wanted here instead would be sg like:
```
if [ -d "${__dir}/.config" ]; then
  echo "Cannot create ${__dir}/.config file as there's already a directory with that name";
  exit 1
fi
touch .config
```

File: examples/upload-media-from-local/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+```
+
+here is the list of arguments we can use:
```
Comment (zoltan-sebestyen-instructure) [inline examples/upload-media-from-local/README.md:15]: I believe this is too verbose, I'd rather have three version of the sample CLI for each case ( single upload, multiple uplod and one of them with the user id parameter)

**Review by zoltan-sebestyen-instructure — CHANGES_REQUESTED** [review]

File: bin/install (hunk 8 above)
Comment (kasra-r77) [reply bin/install:12]: I wanted to check if there is a .config file already or not. so if it is there already we skip it instead of exiting the script and continue with installation of packages. if we add more to this repo and more packages in future and we already have a `.config` file we never can run `bin/install` anymore

File: README.md (hunk 5 above)
Comment (kasra-r77) [reply README.md]: I mentioned it in the parent README since handling authorization is a generic thing but I added a refrence to it here

File: bin/install (hunk 8 above)
Comment (kasra-r77) [reply bin/install:12]: I just checked and found out `touch` command doesn't override the content if the file already exists. I will remove the `if` part for it

File: examples/upload-media-from-local/upload.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import os
+import sys
+
+sys.path.insert(0, os.getcwd())
+from examples.utils.request import handle_request, handle_authorization
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("Host", metavar="host", type=str, help="Host Url")
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply examples/upload-media-from-local/upload.py:11]: done

File: bin/run [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && cd .. && pwd)"
+
+# checking if venv is available
+if [ ! -d "${__dir}/venv" ]; then
+  echo "Virtualenv missing! please run bin/install first"
+  exit 1
+fi
+
+source "${__dir}/venv/bin/activate"
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply bin/run:13]: done

File: bin/install [hunk 12]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+__dir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && cd .. && pwd)"
+
+if [ ! -d "${__dir}/venv" ]; then
+  python3 -m venv "${__dir}/venv"
+fi
+
+source "${__dir}/venv/bin/activate"
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply bin/install:11]: done

File: README.md (hunk 6 above)
Comment (kasra-r77) [reply README.md]: We can! I didn't know about this link, added

File: README.md (hunk 5 above)
Comment (zoltan-sebestyen-instructure) [reply README.md]: Sorry, could you please explain the "parent README" ? This README file is in the root of the repository. I can't see the reference at line #13

File: bin/install (hunk 8 above)
Comment (zoltan-sebestyen-instructure) [reply bin/install]: First, my suggestion actually drops an error if the user accidentally created a directory, so it may have been a better option.
Second, creating an installer only to run `python3 -m venv venv; touch .config` does not really justify its existence. May be it's a better option to put these steps into the README instead, since in this form these scripts only add complexity
What may be an option is to move the sample JSON body from the README to the `bin/install`, so the user don't have
to do it himself.

```
$ bash
$ help test|grep '\-d'
      -d FILE        True if file is a directory.

# and indeed:
$ touch .file
$ mkdir .dir
$ if [ -d .file ]; then echo OK; fi
$ if [ -d .dir ]; then echo OK; fi
OK
```

File: examples/upload-media-from-local/upload.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zoltan-sebestyen-instructure) [reply examples/upload-media-from-local/upload.py:11]: My original suggestion was bad and now it even has a typo :)  may be it could be `help="upload on behalf of a user",`
Now it would look like:
```
❯ bin/run examples/upload-media-from-local/upload.py -h
usage: upload.py [-h] [--user-id USER_ID] host files [files ...]
...
optional arguments:
  -h, --help         show this help message and exit
  --user-id USER_ID  upload on behalf of a user
```

[the remark this answers is not in the document]
Comment (kasra-r77) [reply examples/upload-media-from-local/upload.py:11]: fixed

File: bin/install (hunk 8 above)
Comment (kasra-r77) [reply bin/install]: I dropped the scripts and wrote everything in the README. I wouldn't keep a JSON creation in the installer since as you mentioned it would't justify its existence.  I think README should be okay now 🤔

File: README.md [hunk 13]
```diff
@@ -4,11 +4,40 @@ In this repository, we collect example scripts that can be useful for the team a
...
+{
+  "access_token": "${ACCESS_TOKEN}",
+  "client_id": "${CLIENT_ID}",
+  "client_secret": "${CLIENT_SECRET}",
+  "refresh_token": "${REFRESH_TOKEN}"
+}
+```
+
+Note: You need to authorize yourself through OAuth to get `access_token` and `refresh_token`. if token gets expired, the script will handle fetching a new one. here is how to use [OAuth](https://community.canvaslms.com/t5/The-Product-Blog/Connecting-Studio-OAuth-via-Postman/ba-p/259739)
+
```
Comment (cu2) [inline README.md:29]: Once installed and configured, the user can test it by running:
```
bin/run examples/test --subdomain <school-subdomain>
```

For this we need an examples/test/main.py like this:
```
import argparse

from utils.utils import PublicAPIClient


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--subdomain", type=str, required=True)
    args = parser.parse_args()

    public_api_client = PublicAPIClient(args.subdomain)
    public_api_client.refresh_tokens()
    response = public_api_client.request("get", "ping")

    if response.status_code == 200 and response.text == "pong":
        print("You have successfully configured your public API access.")
    else:
        print("Something is not okay")
        print("status_code:", response.status_code)
        print("response body:", response.text)


if __name__ == "__main__":
    main()
```

File: examples/utils/request.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
+import requests
```
Comment (cu2) [inline examples/utils/request.py:1]: As we discussed, let's
1) move this to `utils/utils.py`
2) use something similar that I sent over in Slack

File: examples/upload-media-from-local/upload.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
+import argparse
```
Comment (cu2) [inline examples/upload-media-from-local/upload.py:1]: As discussed, let's rename to main.py

Here's an updated version that uses the new utils.py I sent over in Slack:

```
import argparse
import os

from utils.utils import PublicAPIClient, request_with_retry


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("files", nargs="+")
    parser.add_argument("--subdomain", type=str, required=True)
    parser.add_argument("--user-id", type=int)
    args = parser.parse_args()

    public_api_client = PublicAPIClient(args.subdomain)
    for media_file in args.files:
        media_filename = os.path.basename(media_file)
        print(f"Uploading {media_filename}")
        with open(media_file, "rb") as f:
            media_id, presigned_url = create_media(public_api_client, args.user_id)
            upload_file(presigned_url, f)
            mark_media_as_uploaded(public_api_client, media_id, media_filename)
        print(f"Uploaded {media_filename}")


def create_media(public_api_client, user_id):
    response = public_api_client.request(
        "post",
        "media/uploads",
        params={"user_id": user_id} if user_id else None,
    )
    if response.status_code != 201:
        raise Exception(f"Could not create media: {response.text}")
    response_body = response.json()
    return (
        response_body["upload"]["id"],
        response_body["upload"]["url"],
    )


def upload_file(presigned_url, f):
    response = request_with_retry(
        "put",
        presigned_url,
        data=f,
    )
    if response.status_code != 200:
        raise Exception(f"Could not upload file: {response.text}")


def mark_media_as_uploaded(public_api_client, media_id, media_filename):
    response = public_api_client.request(
        "post", f"media/uploads/{media_id}/complete", params={"title": media_filename}
    )
    if response.status_code != 200:
        raise Exception(f"Could not mark media as uploaded: {response.text}")


if __name__ == "__main__":
    main()
```

**Review by cu2 — CHANGES_REQUESTED** [review]

File: README.md [hunk 16]
```diff
@@ -4,11 +4,40 @@ In this repository, we collect example scripts that can be useful for the team a
...
+- install dependencies: `pip install -q -r requirements.txt`
+
+### Authorization
+
+In order to authorize the requests:
+
+- create a `.config` file in the root directory
+- add your credentials in a `JSON` format into the `.config` file as following:
+
+```json
```
[the remark this answers is not in the document]
Comment (zoltan-sebestyen-instructure) [reply README.md:19]: Another option could be just have a `.config.sample` file and an instruction to copy it to `.config` at the first time. I.e. we may still not need the `bin/install` :)

File: bin/install [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+{
+    "access_token": "",
+    "client_id": "",
+    "client_secret": "",
+    "refresh_token": ""
+}
+EOT
+fi
+
+
```
Comment (cu2) [inline bin/install:24]: Just a nitpick, but I think one blank line is enough here.

File: README.md [hunk 18]
```diff
@@ -4,22 +4,48 @@ In this repository, we collect example scripts that can be useful for the team a
...
+$ bin/install
+```
+
+This will handle installing the required packages into virtualenv.
+
+### Authorization
+
+In order to authorize the requests, please add your credentials in a `JSON` format into the `config.json` file created by install script.
+
+Note: You need to authorize yourself through OAuth to get `access_token` and `refresh_token`. if token gets expired, the script will handle fetching a new one. here is how to use [OAuth](https://community.canvaslms.com/t5/The-Product-Blog/Connecting-Studio-OAuth-via-Postman/ba-p/259739)
```
Comment (cu2) [inline README.md:19]: I would rather say this:

You can generate the necessary credentials as described [here](https://community.canvaslms.com/t5/The-Product-Blog/Connecting-Studio-OAuth-via-Postman/ba-p/259739).

Once you have them in config.json, you won't have to deal with them anymore: the example scripts automatically handle renewing expired tokens.

File: README.md [hunk 19]
```diff
@@ -4,22 +4,48 @@ In this repository, we collect example scripts that can be useful for the team a
...
+
+### Testing
+
+In order to evaluate the authorization, please run:
```
Comment (cu2) [inline README.md:23]: In order to confirm your credentials, please run:

File: examples/upload-media-from-local/README.md [hunk 20]
```diff
@@ -0,0 +1,35 @@
+# Upload media from local
+
+This is a script to handle uploading media locally to Studio.
```
Comment (cu2) [inline examples/upload-media-from-local/README.md:3]: I would rather say: This is a script that can upload media files to Studio from local.

File: examples/upload-media-from-local/README.md [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+## How to use
+
+Before starting using this script don't forget to authorize your request. this is explained here: [Authorization](../../README.md#authorization)
+
+For using this script we should run it through runner script in `bin` folder:
```
Comment (cu2) [inline examples/upload-media-from-local/README.md:9]: I would remove this and also listing arguments. The former is explained in the root README, the latter is handled by argparse's help.

I would just have three examples:

To upload a single file:

```bash
bin/run examples/upload-media-from-local --subdomain <school_subdomain> /Users/some_user/Desktop/media/file.mp4
```

To upload multiple files:

```bash
bin/run examples/upload-media-from-local --subdomain <school_subdomain> /Users/some_user/Desktop/media/*
```

To upload on behalf of another user:
```bash
bin/run examples/upload-media-from-local --subdomain <school_subdomain> --user-id 3 /Users/some_user/Desktop/media/file.mp4
```

**Review by cu2 — CHANGES_REQUESTED** [review]
Getting there, just a few minor changes needed.

**Comment by cu2** [comment]
Also, don't forget to chmod +x your bin scripts. Now they are not executable.

File: README.md [hunk 22]
```diff
@@ -4,11 +4,40 @@ In this repository, we collect example scripts that can be useful for the team a
 
 ## Prerequisites
 
-TBD
+- install a python 3 interpreter (if not installed already)
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md]: done

File: README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md]: Andras mentioned that this is discussed and you agreed to keep these

File: examples/utils/request.py (hunk 14 above)
Comment (kasra-r77) [reply examples/utils/request.py]: done

File: examples/upload-media-from-local/upload.py (hunk 15 above)
Comment (kasra-r77) [reply examples/upload-media-from-local/upload.py]: done

File: bin/install (hunk 17 above)
Comment (kasra-r77) [reply bin/install]: removed

File: config.json [hunk 23]
```diff
@@ -0,0 +1 @@
+{"access_token": "N2eaG3vq0apme9pYbtJzZmBQb4elRa5a-5Vj42b9Mv0", "refresh_token": "oljxtVeiM7JKK7hGEBURJTgpc4Ffqod1EtTd96Sagzg", "client_id": "dXitp230X7M1s-WNCPlusc3-ehrpwiLnXBhTweaPSC8", "client_secret": "k5_7s2sQr_PgWcuwS1f8_nt3Js_XJRIedeJnCXa8AGk"}
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply config.json]: it is okay, those were some local creds and not valid anymore

File: README.md [hunk 24]
```diff
@@ -4,22 +4,48 @@ In this repository, we collect example scripts that can be useful for the team a
...
+
+```bash
+$ bin/install
+```
+
+This will handle installing the required packages into virtualenv.
+
+### Authorization
+
+In order to authorize the requests, please add your credentials in a `JSON` format into the `config.json` file created by install script.
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md]: done

File: README.md (hunk 18 above)
Comment (kasra-r77) [reply README.md]: done

File: README.md [hunk 25]
```diff
@@ -4,22 +4,48 @@ In this repository, we collect example scripts that can be useful for the team a
 
 ## Prerequisites
 
-TBD
+To run python scripts inside `examples` directory you should first run the installer script:
+
+```bash
+$ bin/install
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md:10]: I wanted to show it bash-ish. I removed them

File: README.md [hunk 26]
```diff
@@ -4,22 +4,48 @@ In this repository, we collect example scripts that can be useful for the team a
...
+```
 
 ## Language
 
-The scripts in this repository are prefered to be written in Python as it's widely adopted and easy to read.
+The scripts in this repository are preferred to be written in Python as it's widely adopted and easy to read.
 
 ## Layout of the repository
 
 Each script needs to have its own subdirectory under the `examples` directory and needs to have a `README.md` explaining its purpose and how to use it.
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md:47]: done

File: examples/upload-media-from-local/README.md (hunk 20 above)
Comment (kasra-r77) [reply examples/upload-media-from-local/README.md]: done

File: examples/upload-media-from-local/README.md [hunk 27]
```diff
@@ -0,0 +1,35 @@
+# Upload media from local
+
+This is a script to handle uploading media locally to Studio.
+
+## How to use
+
+Before starting using this script don't forget to authorize your request. this is explained here: [Authorization](../../README.md#authorization)
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply examples/upload-media-from-local/README.md]: done

File: examples/upload-media-from-local/README.md (hunk 21 above)
Comment (kasra-r77) [reply examples/upload-media-from-local/README.md]: done

File: .gitignore [hunk 28]
```diff
@@ -0,0 +1,5 @@
+__pycache__/
+venv/
+.config
```
[the remark this answers is not in the document]
Comment (cu2) [reply .gitignore:3]: Cool. 😄  But please remove .config from .gitignore.

File: README.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (cu2) [reply README.md:10]: Except from here. 😄

File: README.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (cu2) [reply README.md:47]: Sorry, I was using bad English:
- If it *is* written in Python
- and I also think the comma after "assumes" it not needed

File: README.md [hunk 29]
```diff
@@ -4,22 +4,52 @@ In this repository, we collect example scripts that can be useful for the team a
...
+$ bin/install
+```
+
+This will handle installing the required packages into virtualenv.
+
+### Authorization
+
+In order to authorize the requests, please add your credentials in `JSON` format into the `config.json` file created by install script.
+
+Note: You can generate the necessary credentials as described [here](https://community.canvaslms.com/t5/The-Product-Blog/Connecting-Studio-OAuth-via-Postman/ba-p/259739)
```
Comment (cu2) [inline README.md:19]: "Note:" is not needed I think.

File: examples/upload-media-from-local/README.md [hunk 30]
```diff
@@ -0,0 +1,29 @@
+# Upload media from local
+
+This is a script that can upload media files to Studio from local.
+
+## How to use
+
+Before using this script, don't forget to configure your OAuth credentials (see [here](../../README.md#authorization)).
+
+Once you have them in config.json, you won't have to deal with them anymore: the example scripts automatically handle renewing expired tokens.
```
Comment (cu2) [inline examples/upload-media-from-local/README.md:9]: I would delete this paragraph: it's a detail that is described in the root README, I wouldn't add it to anywhere else.

File: examples/upload-media-from-local/README.md [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+
+## examples
```
Comment (cu2) [inline examples/upload-media-from-local/README.md:11]: Let's remove this heading: these are all parts of "How to use"

File: utils/utils.py [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+    def _save_config(self):
+        with open(self.config_path, "wt") as f:
+            json.dump(self.config, f, indent=4, sort_keys=True)
+
+
+def request_with_retry(method, url, headers=None, params=None, data=None, retry=3):
+    response = getattr(requests, method)(url, headers=headers, params=params, data=data)
+    if response.status_code >= 500:
+        if retry > 0:
+            response = request_with_retry(
```
Comment (cu2) [inline utils/utils.py:76]: Let's add
```
time.sleep(3)
```
here, so the retry is not that agressive.

File: utils/utils.py [hunk 33]
```diff
@@ -0,0 +1,84 @@
+import json
+import os
+import requests
+
```
Comment (cu2) [inline utils/utils.py:4]: Let's add
```
import time
```
here, so we can use the sleep method below.

**Review by cu2 — CHANGES_REQUESTED** [review]
There's a `.keep` file in the `examples` folder, which is not necessary anymore now that we have context in the folder. Please remove it.

Also, let's add a README.md to examples/test with something like this:

# Test API access

This script tests renewing and using OAuth tokens. It can be used for confirming that the credentials in `config.json` are valid.

## How to use

```bash
bin/run examples/test --subdomain <school_subdomain>
```

File: utils/utils.py (hunk 33 above)
Comment (kasra-r77) [reply utils/utils.py:5]: done

File: examples/upload-media-from-local/README.md (hunk 30 above)
Comment (kasra-r77) [reply examples/upload-media-from-local/README.md]: done

File: README.md (hunk 29 above)
Comment (kasra-r77) [reply README.md]: done

File: README.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md:47]: done

**Review by cu2 — DISMISSED** [review]

File: bin/install (hunk 8 above)
Comment (zoltan-sebestyen-instructure) [reply bin/install]: Let's resolve it, since `config.json` is unlikely to be created accidentally as a directory, also `touch` is gone.

File: examples/test/main.py [hunk 34]
```diff
@@ -0,0 +1,24 @@
...
+    args = parser.parse_args()
+
+    public_api_client = PublicAPIClient(args.subdomain)
+    public_api_client.refresh_tokens()
+    response = public_api_client.request("get", "ping")
+
+    if response.status_code == 200 and response.text == "pong":
+        print("You have successfully configured your public API access.")
+    else:
+        print("Something is not okay")
```
Comment (zoltan-sebestyen-instructure) [inline examples/test/main.py:18]: Something is not too specific. I would prefer "Error happened during public API access test" or similar.

**Review by zoltan-sebestyen-instructure — CHANGES_REQUESTED** [review]

File: README.md [hunk 35]
```diff
@@ -4,22 +4,52 @@ In this repository, we collect example scripts that can be useful for the team a
...
+```bash
+bin/run <path_to_script_folder> <arguments>
+```
 
 ## Language
 
-The scripts in this repository are prefered to be written in Python as it's widely adopted and easy to read.
+The scripts in this repository are preferred to be written in Python as it's widely adopted and easy to read.
+
+If it is written in Python, name it `main.py`, because the runner script assumes it's named like that.
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md:43]: We agreed that the language and layout section will be moved to confluence page

File: bin/run [hunk 36]
```diff
@@ -0,0 +1,22 @@
...
+
+if [[ $# -lt 1 ]]; then
+  echo "Usage: bin/run <script_path>"
+  exit 1
+fi
+
+script_path="${1}"
+shift
+
+PYTHONPATH="${__dir}" exec python "${script_path}/main.py" "$@"
```
[the remark this answers is not in the document]
Comment (kasra-r77) [reply bin/run:22]: We agreed to add the file as well.

File: examples/test/main.py (hunk 34 above)
Comment (kasra-r77) [reply examples/test/main.py]: done

File: bin/run (hunk 36 above)
[the remark this answers is not in the document]
Comment (kasra-r77) [reply bin/run]: done

File: README.md (hunk 35 above)
[the remark this answers is not in the document]
Comment (kasra-r77) [reply README.md]: done moved to CF

**Review by cu2 — APPROVED** [review]

File: README.md (hunk 5 above)
Comment (zoltan-sebestyen-instructure) [reply README.md]: @kasra-r77 any chance to fix that?

File: examples/upload-media-from-local/upload.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (zoltan-sebestyen-instructure) [reply examples/upload-media-from-local/upload.py]: Sorry, but I still can't see the help comment suggested above:
```
    parser.add_argument("--user-id", type=int, required=False
                       help="Optional parameter to upload on behalf a user")
```

File: README.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (zoltan-sebestyen-instructure) [reply README.md]: it's done, so I resolve it.

File: examples/upload-media-from-local/upload.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kasra-r77) [reply examples/upload-media-from-local/upload.py]: done

File: README.md (hunk 5 above)
Comment (kasra-r77) [reply README.md]: done

**Review by zoltan-sebestyen-instructure — APPROVED** [review]

File: examples/upload-media-from-local/main.py [hunk 37]
```diff
@@ -0,0 +1,61 @@
+import argparse
+import os
+
+from utils.utils import PublicAPIClient, request_with_retry
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("files", nargs="+", help="path to the media file/folder")
```
Comment (cu2) [inline examples/upload-media-from-local/main.py:9]: Sorry for nitpicking, but it's not "file/folder", but "file(s)". If you pass a folder, it won't work, the example passes a list of files (the shell expands the folder/* to a list of files).

**Review by cu2 — CHANGES_REQUESTED** [review]

File: examples/upload-media-from-local/main.py (hunk 37 above)
Comment (kasra-r77) [reply examples/upload-media-from-local/main.py]: yeah I agree this is a better explanation for it. done

**Review by cu2 — APPROVED** [review]

**Review by zoltan-sebestyen-instructure — APPROVED** [review]

**Comment by cu2** [comment]
I QA'd this by running both the test and the upload script against my staging Sandbox: they work as expected. I also tested the install script (by deleting my venv and config.json and running it).
